Packs/CrowdStrikeFalcon/Integrations/CrowdStrikeFalcon/CrowdStrikeFalcon.py

```diff
@@ -1110,70 +1156,137 @@ def fetch_incidents():
+                raw_res['type'] = "detections"
+                for detection in raw_res.get("resources"):
+                    incident = detection_to_incident(detection)
+                    incident_date = incident['occurred']
+
+                    incident_date_timestamp = int(parse(incident_date).timestamp() * 1000)
+
+                    # make sure that the two timestamps are in the same length
+                    if len(str(incident_date_timestamp)) != len(str(last_fetch_timestamp)):
+                        incident_date_timestamp, last_fetch_timestamp = timestamp_length_equalization(
+                            incident_date_timestamp, last_fetch_timestamp)
+
+                    # Update last run and add incident if the incident is newer than last fetch
+                    if incident_date_timestamp > last_fetch_timestamp:
+                        last_fetch = incident_date
+                        last_fetch_timestamp = incident_date_timestamp
+                        last_detection_id = json.loads(incident['rawJSON']).get('detection_id')
+
+                    incidents.append(incident)
+
+            demisto.setLastRun({'first_behavior_time': last_fetch, 'last_detection_id': last_detection_id})
+
+    if 'incidents' in incidents_or_detections:
+
+        last_run = demisto.getLastRun()
+
+        last_fetch = last_run.get('first_behavior_incident_time')
+
+        if last_fetch is None:
+            last_fetch, _ = parse_date_range(FETCH_TIME, date_format='%Y-%m-%dT%H:%M:%SZ')
+
+        last_fetch_timestamp = int(parse(last_fetch).timestamp() * 1000)
+
+        last_incident_id = str(last_run.get('last_incident_id'))
+
+        fetch_query = demisto.params().get('fetch_query')
+
+        if fetch_query:
+            fetch_query = "modified_timestamp:>'{time}'+{query}".format(time=last_fetch, query=fetch_query)
+            incidents_ids = demisto.get(get_fetch_incidents(filter_arg=fetch_query), 'resources')
 
-        # if the first detection in this pull is different than the last detection fetched we bring it as well
         else:
-            first_index_to_fetch = 0
+            incidents_ids = demisto.get(get_fetch_incidents(last_created_timestamp=last_fetch), 'resources')
 
-        # Limit the results to INCIDENTS_PER_FETCH`z
-        last_index_to_fetch = INCIDENTS_PER_FETCH + first_index_to_fetch
-        detections_ids = detections_ids[first_index_to_fetch:last_index_to_fetch]
-        raw_res = get_detections_entities(detections_ids)
+        if incidents_ids:
+
+            # make sure we do not fetch the same detection again.
+            if last_incident_id == incidents_ids[0]:
+                first_index_to_fetch = 1
+
+                # if this is the only detection - dont fetch.
+                if len(incidents_ids) == 1:
+                    return incidents
+
+            # if the first detection in this pull is different than the last detection fetched we bring it as well
+            else:
```

Comment:
A lot of the code here and down the line seems to be duplicate.
Let's extract any duplicate logic to external functions